Add strikethrough and underline formatter on text schema

necessarylion opened the pull request:
- Work on different font sizes
- Work on multiple lines
- Work on text with character spacing
- Work on text with align/vertical align

Designer

<img width="1390" alt="Screenshot 2024-07-12 at 11 25 38 AM" src="https://github.com/user-attachments/assets/9d5da64d-4c4d-4ce1-9e19-9b32f1df5f6c"> <br>


Generated PDF

<img width="1390" alt="Screenshot 2024-07-12 at 11 25 54 AM" src="https://github.com/user-attachments/assets/f4d9f00b-04f6-4283-91ef-98e44c249180">

peteward commented:
Hi @necessarylion,

This is very cool, thanks for the contribution.

I think it's nearly all working except for character spacing is missing from the line length calculations:

![Screenshot 2024-07-13 at 11 45 19](https://github.com/user-attachments/assets/66935fa7-558b-40e7-b255-b7e95b3eaf95)
![Screenshot 2024-07-13 at 11 45 42](https://github.com/user-attachments/assets/0ee8b72f-7093-4b3a-88a8-46e56e5d42b8)

But everything else that I've checked so far works.

I think for the UI panel it would be good to add the icons into an existing row, rather than adding a new one, perhaps on the left hand side of the Alignment options with a heading `Formatting`? We can make the other options thinner, or perhaps make text alignment buttons too then they would use less horizontal space. one for @hand-dot to comment on.

Also, I think it would be good to add an integration test for this formatting. I would be happy to help add this though after it is merged.

necessarylion commented:
Hi @peteward, thank you for the review. I have fixed character spacing issue [here](https://github.com/pdfme/pdfme/pull/494/commits/da9e5c65ac78f0ada2aec5e26f25358e843ac79c). 

Regarding the UI panel, alignment components were utilized across various schemas, whereas the new strikethrough and underline features are specific to the text schema. This is why I introduced them as new components. My initial idea was to place them next to the other alignment options, but I couldn't do so because they are shared components and also created as the custom widget for [form-render](https://www.npmjs.com/package/form-render).

If you have a better suggestion, please let me know; I'm happy to consider it.

peteward commented:
Thanks for the fast update @necessarylion .

I meant next to the _text_ alignment options (Horizontal: left, center, right & Vertical: bottom, middle, top), not the widget alignment options.

I might play around with this later today and see how it looks 👍

peteward commented on packages/ui/src/components/Designer/RightSidebar/DetailView/index.tsx:
```diff
@@ -35,6 +36,7 @@ const DetailView = (
   useEffect(() => {
     const newWidgets: typeof widgets = {
       AlignWidget: (p) => <AlignWidget {...p} {...props} options={options} />,
+      TextFormatWidget: (p) => <TextFormatWidget {...p} {...props} options={options} />,
```
I can see what you have done here and why, but I think this needs to be more generic, and not tied to an implementation just for the text schema.

I think we could have a `BooleanButtonsWidget` which could be generic enough to pass in the names of attributes and SVG details to be rendered.

peteward replied:
Using a quick test I could make this:

![Screenshot 2024-07-16 at 06 32 12](https://github.com/user-attachments/assets/01fe2b87-cd39-46d2-a253-02f8ee004a80)

I think we could do a similar buttons widget for text alignment too. I will try this later today.

necessarylion replied:
<img width="394" alt="Screenshot 2024-07-16 at 4 18 34 PM" src="https://github.com/user-attachments/assets/718243c6-032e-49e2-a85a-0a994659f74d">

@peteward How about like this?

And we can configure buttons like below

```ts
export enum Formatter {
  STRIKETHROUGH = 'strikethrough',
  UNDERLINE = 'underline',
  ALIGNMENT = 'alignment',
  VERTICAL_ALIGNMENT = 'verticalAlignment',
}


{
    title: i18n('schemas.text.format'),
    widget: 'ButtonGroup',
    buttons: [
      { key: Formatter.STRIKETHROUGH, svg: TextStrikethroughIcon, type: 'boolean' },
      { key: Formatter.UNDERLINE, svg: TextUnderlineIcon, type: 'boolean' },
      { key: Formatter.ALIGNMENT, svg: TextAlignLeftIcon, type: 'select', value: DEFAULT_ALIGNMENT },
      { key: Formatter.ALIGNMENT, svg: TextAlignCenterIcon, type: 'select', value: ALIGN_CENTER},
      { key: Formatter.ALIGNMENT, svg: TextAlignRightIcon, type: 'select', value: ALIGN_RIGHT},
      { key: Formatter.VERTICAL_ALIGNMENT, svg: TextVerticalAlignTopIcon, type: 'select', value: DEFAULT_VERTICAL_ALIGNMENT },
      { key: Formatter.VERTICAL_ALIGNMENT, svg: TextVerticalAlignMiddleIcon, type: 'select', value: VERTICAL_ALIGN_MIDDLE},
      { key: Formatter.VERTICAL_ALIGNMENT, svg: TextVerticalAlignBottomIcon, type: 'select', value: VERTICAL_ALIGN_BOTTOM},
    ]
    span: 16,
 };
```

necessarylion replied:
I will push the code, so that you can check. And please feel free to give the feedback.

peteward replied:
this is excellent, I really like it!

I had just played around with something similar but splitting the buttons into groups (but it was hard to fit one one line)

![Screenshot 2024-07-16 at 10 51 09](https://github.com/user-attachments/assets/295e0835-29c3-4b8c-b795-c64d7883dce4)


Your code is better though and would give the flexibility to split them if desired 👍 

I think one final improvement would be to use SVGs that are more familiar to users working with text, like:

![Screenshot 2024-07-16 at 10 54 38](https://github.com/user-attachments/assets/95eb5114-deae-42b1-9b92-3b4a13383adb)
![Screenshot 2024-07-16 at 10 55 31](https://github.com/user-attachments/assets/909b7f2c-3fa2-45c0-8151-56025b970fe1)

We could always add these later though.

@hand-dot, what do you think of this PR?

necessarylion replied:
> I think one final improvement would be to use SVGs that are more familiar to users working with text.

Yeah agreed, but I was just following the same icon from upper align icons.

peteward replied:
I understand.

Personally, I think it would look clearer to have different icons from the widget alignment ones to distinguish the functionality.

necessarylion replied:
<img width="430" alt="Screenshot 2024-07-16 at 8 19 01 PM" src="https://github.com/user-attachments/assets/d7e6c0d2-ea4b-40e4-8589-54451446b3ba">

Updated.

peteward replied:
great work!

hand-dot replied:
yes yes! Really nice.

peteward commented on packages/common/src/schema.ts:
```diff
@@ -56,6 +56,7 @@ export const Dict = z.object({
   'schemas.text.max': z.string(),
   'schemas.text.fit': z.string(),
   'schemas.text.dynamicFontSize': z.string(),
+  'schemas.text.format': z.string(),
```
do we still need this?

necessarylion replied:
Yes. It is for the [schema title](https://github.com/pdfme/pdfme/pull/494/files#diff-84fadaad049676d98756eacc6206f5f695bca78fe837522dd9b68bca10cb73a6R71-R72).

peteward replied:
ah sorry, yes this is translations not schema definition 👍

hand-dot commented:
@necessarylion 
I will find time to review and test as soon as possible. Please wait a moment.
Thank you for the excellent work.

@peteward 
I'm also very pleased with the improvements from a UX perspective.

hand-dot commented:
@necessarylion 
This feature is also backward compatible, so it's scheduled for release in the next version.
Thank you for your excellent work!!

@peteward 
If you have any additional comments, please let me know. If not, let's merge this.

peteward reviewed (approved):
Great!
